File: apps/frontend/lib/api/predictions.ts
/**
 * Prediction Market API Client
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

export interface PredictionMarket {
  id: string;
  question: string;
  status: 'ACTIVE' | 'CLOSED' | 'RESOLVED';
  winner: 'YES' | 'NO' | null;
  yes_pool: number;
  no_pool: number;
  total_bets: number;
  start_date: string;
  end_date: string;
  category: string;
  icon: string;
  total_pool: number;
  yes_percent: number;
  no_percent: number;
}

export interface VoteRequest {
  market_id: string;
  choice: 'YES' | 'NO';
  amount: number;
  wallet?: string;
}

export interface VoteResponse {
  success: boolean;
  vote_id: string;
  market_id: string;
  choice: 'YES' | 'NO';
  amount: number;
  new_yes_pool: number;
  new_no_pool: number;
  yes_percent: number;
  no_percent: number;
}

export interface MarketStats {
  market_id: string;
  question: string;
  status: string;
  yes_pool: number;
  no_pool: number;
  total_pool: number;
  yes_percent: number;
  no_percent: number;
  total_bets: number;
  days_remaining: number;
  ends_in: string;
}

/**
 * Fetch all prediction markets
 */
export async function getPredictionMarkets(): Promise<PredictionMarket[]> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/predictions/markets`);

    if (!response.ok) {
      throw new Error(`Failed to fetch markets: ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Error fetching prediction markets:', error);
    return [];
  }
}

/**
 * Fetch a specific prediction market
 */
export async function getPredictionMarket(marketId: string): Promise<PredictionMarket | null> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/predictions/markets/${marketId}`);

    if (!response.ok) {
      if (response.status === 404) {
        return null;
      }
      throw new Error(`Failed to fetch market: ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error(`Error fetching market ${marketId}:`, error);
    return null;
  }
}

/**
 * Place a YES/NO vote on a prediction market
 */
export async function placeVote(vote: VoteRequest): Promise<VoteResponse> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/predictions/vote`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(vote),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to place vote');
    }

    return await response.json();
  } catch (error) {
    console.error('Error placing vote:', error);
    throw error;
  }
}

/**
 * Get market statistics
 */
export async function getMarketStats(marketId: string): Promise<MarketStats | null> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/predictions/markets/${marketId}/stats`);

    if (!response.ok) {
      throw new Error(`Failed to fetch market stats: ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error(`Error fetching market stats for ${marketId}:`, error);
    return null;
  }
}

File: apps/backend/app/api/v1/predictions.py
from fastapi import APIRouter, HTTPException
from app.models.prediction import PredictionMarket, VoteRequest, VoteResponse
from typing import List
from datetime import datetime, timedelta
import random
import uuid

router = APIRouter()

# In-memory storage for demo (replace with database in production)
MARKETS_DB = {}
VOTES_DB = []

# Initialize the Eden Haus hackathon market
HACKATHON_MARKET = PredictionMarket(
    id="eden-haus-hackathon",
    question="Will Eden Haus win the Cronos x402 Hackathon?",
    status="ACTIVE",
    yes_pool=525.0,  # Starting with some demo data
    no_pool=725.0,
    total_bets=0,
    start_date=datetime.utcnow() - timedelta(days=1),
    end_date=datetime.utcnow() + timedelta(days=31),
    category="Crypto",
    icon="🏆"
)

MARKETS_DB["eden-haus-hackathon"] = HACKATHON_MARKET

@router.get("/markets", response_model=List[PredictionMarket])
async def get_markets():
    """Get all prediction markets"""
    return list(MARKETS_DB.values())

@router.get("/markets/{market_id}", response_model=PredictionMarket)
async def get_market(market_id: str):
    """Get a specific prediction market by ID"""
    market = MARKETS_DB.get(market_id)
    if not market:
        raise HTTPException(status_code=404, detail="Market not found")
    return market

@router.post("/vote", response_model=VoteResponse)
async def place_vote(vote: VoteRequest):
    """Place a YES/NO vote on a prediction market"""

    # Validate market exists
    market = MARKETS_DB.get(vote.market_id)
    if not market:
        raise HTTPException(status_code=404, detail="Market not found")

    # Validate market is active
    if market.status != "ACTIVE":
        raise HTTPException(status_code=400, detail="Market is not active")

    # Validate market hasn't ended
    if datetime.utcnow() > market.end_date:
        raise HTTPException(status_code=400, detail="Market has ended")

    # Update pools
    if vote.choice == "YES":
        market.yes_pool += vote.amount
    else:
        market.no_pool += vote.amount

    market.total_bets += 1

    # Store vote
    vote_id = str(uuid.uuid4())
    VOTES_DB.append({
        "vote_id": vote_id,
        "market_id": vote.market_id,
        "choice": vote.choice,
        "amount": vote.amount,
        "wallet": vote.wallet,
        "timestamp": datetime.utcnow()
    })

    # Update market in DB
    MARKETS_DB[vote.market_id] = market

    return VoteResponse(
        success=True,
        vote_id=vote_id,
        market_id=vote.market_id,
        choice=vote.choice,
        amount=vote.amount,
        new_yes_pool=market.yes_pool,
        new_no_pool=market.no_pool,
        yes_percent=market.yes_percent,
        no_percent=market.no_percent
    )

@router.get("/markets/{market_id}/stats")
async def get_market_stats(market_id: str):
    """Get real-time statistics for a market"""
    market = MARKETS_DB.get(market_id)
    if not market:
        raise HTTPException(status_code=404, detail="Market not found")

    # Calculate time remaining
    now = datetime.utcnow()
    time_remaining = (market.end_date - now).total_seconds()
    days_remaining = max(0, int(time_remaining / 86400))

    return {
        "market_id": market.id,
        "question": market.question,
        "status": market.status,
        "yes_pool": market.yes_pool,
        "no_pool": market.no_pool,
        "total_pool": market.total_pool,
        "yes_percent": market.yes_percent,
        "no_percent": market.no_percent,
        "total_bets": market.total_bets,
        "days_remaining": days_remaining,
        "ends_in": f"{days_remaining} days"
    }
